### PR TITLE
bpo-34656: Avoid relying on signed overflow in _pickle memos.

### PR DESCRIPTION
**Pull request by benjaminp**
https://bugs.python.org/issue1621

### CODE REVIEW

File: Modules/_pickle.c [hunk 1]
```diff
@@ -7061,7 +7057,7 @@ Unpickler_set_memo(UnpicklerObject *self, PyObject *obj)
   error:
     if (new_memo_size) {
         i = new_memo_size;
```
Comment (sir-sigurd) [inline Modules/_pickle.c]: It looks like this line is redundant now.

File: Modules/_pickle.c [hunk 2]
```diff
@@ -821,22 +821,18 @@ _PyMemoTable_Lookup(PyMemoTable *self, PyObject *key)
...
     PyMemoEntry *oldentry, *newentry;
-    Py_ssize_t new_size = MT_MINSIZE;
+    size_t new_size = MT_MINSIZE;
     Py_ssize_t to_process;
 
     assert(min_size > 0);
 
     /* Find the smallest valid table size >= min_size. */
-    while (new_size < min_size && new_size > 0)
+    while (new_size < min_size)
```
Comment (serhiy-storchaka) [inline Modules/_pickle.c]: Wouldn't this cause an infinite loop if `min_size` > `PY_SSIZE_T_MAX`?

Comment (benjaminp) [reply Modules/_pickle.c]: Yes, I will fix.

File: Modules/_pickle.c [hunk 3]
```diff
@@ -821,22 +820,23 @@ _PyMemoTable_Lookup(PyMemoTable *self, PyObject *key)
...
-    while (new_size < min_size && new_size > 0)
-        new_size <<= 1;
-    if (new_size <= 0) {
+    if (min_size > PY_SSIZE_T_MAX) {
         PyErr_NoMemory();
         return -1;
     }
+
+    /* Find the smallest valid table size >= min_size. */
+    while (new_size < min_size)
```
Comment (sir-sigurd) [inline Modules/_pickle.c]: Braces are required here per PEP 7.

File: Modules/_pickle.c [hunk 4]
```diff
@@ -909,7 +909,8 @@ PyMemoTable_Set(PyMemoTable *self, PyObject *key, Py_ssize_t value)
      * Very large memo tables (over 50K items) use doubling instead.
      * This may help applications with severe memory constraints.
      */
-    if (!(self->mt_used * 3 >= (self->mt_mask + 1) * 2))
+    size_t triple_used = self->mt_used * 3;
+    if (triple_used > self->mt_used && triple_used < self->mt_allocated * 2)
```
Comment (sir-sigurd) [inline Modules/_pickle.c]: > triple_used > self->mt_used

Is this a wrapping/overflow detection?

Comment (benjaminp) [reply Modules/_pickle.c]: yes

**Pull request merged**